### DISCUSSION TITLE
Out of order drawing for 3D / Layered 3D rendering?

### DISCUSSION THREAD

**Discussion by KnockerPulsar** [raysan5/raylib#2373] [Q&A] [answered]
Hello. I'm trying to build a crude 3D editor, I've been trying to figure out a way to render handles (the little thingys you move objects around with) in front of other geometry. Is there a way to render 3D geometry out of order or into different layers? 

Note: I'm currently using raylib's drawing functions (DrawCube, DrawLine3D, etc..). I would like to work with meshes later on but a solution that works for these objects would be much appreciated too.

What I've tried so far:
- Disabling the depth mask/test -> Applies for all objects, 3D rendering order completely broken.
- Rendering to a texture then rendering that texture right before rendering the frame -> No success so far.
- Transforming the handle ends to screen space coordinates, then using `DrawLine` -> Lines point the wrong way, besides that's only a band aid over a bigger issue.
- Having two rendering passes with the second one not clearing -> Performance drop + didn't work

**Comment by bosoni** [comment]
Are you rendering big Model at once (which contains meshes)? 
Because you are the one who make these layers and sorting lists (if not using just big Model).

raylib renders all meshes (from Model) on one drawcall (check   rmodels.c  -> void DrawModelEx() -function ). Copy that function and modify it to work as you wish. 

Or load every mesh as one model, and then render them like you wish (then you can control depth and other tests). Get mesh's bounding box, its center, mesh's rotation and then render lines.  ...some ideas, hopefully got some ideas.

**Reply by KnockerPulsar** [reply]
Currently, I'm just using raylib's drawing functions (`DrawCube`, `DrawLine3D`, etc..) . Not sure if that would change anything. I also checked out `DrawModelEx` but it seems way over my head.

**Reply by KnockerPulsar** [reply]
Doing a quick google search, I stumbled across #682. I'll try this a bit more when I have the time and will report the results here.

**Comment by KnockerPulsar** [comment, marked as answer]
So, I **think** I figured it out following raysan's suggestions in #682. What you need is:
1. Create two (or more) render textures, I think normal textures are fine, but that's what I did.
2. Wrap every region you want to draw with your texture with the `BeginX()` and `EndX()` functions you currently use.
You can clear the background with whatever color you want.
```cpp
BeginTextureMode(bg);
ClearBackground(RAYWHITE);
BeginMode3D(camera);

// DrawCube() or any other 3D drawing function

EndMode3D();
EndTextureMode();
```

3. For the foreground, you should do the same as step 2, but clear the background with a color with alpha equal to zero.
You can play with the alpha to get cool effects like dimming the background for example.

```cpp
BeginTextureMode(fg);
BeginMode3D(camera);
ClearBackground(Fade(WHITE, 0.0f));

// Functions to draw some foreground 3D objects

EndMode3D();
EndTextureMode();
```
4. After all layers are drawn to their proper textures, you can now draw the render textures back to front. Make sure the order is correct because we draw the background with an opaque color, so it will cover up whatever's behind it.
You might notice that I'm using `-screenHeight`, If you remove the negative sign you'll find out that the textures are actually vertically flipped (3D rendering shenanigans I guess).

```cpp
BeginDrawing();

DrawTextureRec(bg.texture, {0, 0, screenWidth, -screenHeight}, {0, 0},  WHITE);
DrawTextureRec(fg.texture, {0, 0, screenWidth, -screenHeight}, {0, 0},   WHITE);

// UI/Text drawing functions

EndDrawing();
```

If you're trying to do this for UI, you can check out [raygui](https://github.com/raysan5/raygui/) or [rlImGui](https://github.com/raylib-extras/rlImGui). [Here's](https://github.com/KnockerPulsar/raytracer-editor/commit/61b461aab7f907a0c38c76215dd5718ae3e5ad8d) the commit when I added rlImGui and layers.
